crates/turborepo-auth/src/lib.rs

```diff
@@ -12,19 +12,139 @@ mod ui;
 pub use auth::*;
 pub use error::Error;
 pub use login_server::*;
+use turborepo_api_client::TokenClient;
+use turborepo_vercel_api::token::ResponseTokenMetadata;
 
 /// Token is the result of a successful login. It contains the token string and
-/// a boolean indicating whether the token already existed on the filesystem.
-#[derive(Debug)]
+/// potentially metadata about the token.
+#[derive(Debug, Clone)]
 pub enum Token {
+    /// An existing token on the filesystem
     Existing(String),
+    /// A token that was just created, but not yet written to the filesystem
     New(String),
 }
 impl Token {
-    pub fn into_inner(self) -> String {
+    pub fn new(token: String) -> Self {
+        Self::New(token)
+    }
+    pub fn existing(token: String) -> Self {
+        Self::Existing(token)
+    }
+    /// Checks if the token is valid. We do a few checks:
+    /// 1. Fetch the token metadata.
+    /// 2. From the metadata, check if the token is active.
+    /// 3. If the token is a SAML SSO token, check if it's expired.
+    pub async fn is_valid(&self, client: &impl TokenClient) -> Result<bool, Error> {
+        let metadata = self.fetch_metadata(client).await?;
+        let active = is_token_active(&metadata);
+        Ok(active)
+    }
+
+    async fn fetch_metadata(
+        &self,
+        client: &impl TokenClient,
+    ) -> Result<ResponseTokenMetadata, Error> {
+        client
+            .get_metadata(self.into_inner())
+            .await
+            .map_err(Error::from)
+    }
+    /// Returns the underlying token string.
+    pub fn into_inner(&self) -> &str {
         match self {
-            Self::Existing(s) => s,
-            Self::New(s) => s,
+            Self::Existing(token) | Self::New(token) => token.as_str(),
+        }
+    }
+}
+
+fn current_unix_time() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_millis()
+}
+
+// As of the time of writing, this should always be true, since a token that
+// isn't active returns an error when fetching metadata for the token.
+fn is_token_active(metadata: &ResponseTokenMetadata) -> bool {
+    let current_time = current_unix_time();
+    let active_at = metadata.active_at;
+
+    let active_less_than_current = active_at <= current_time;
+    let all_scopes_active = metadata.scopes.iter().all(|scope| {
+        scope
+            .expires_at
+            .map_or(true, |expires_at| expires_at > current_time)
+    });
```

Comment:
Complete bikeshed. My brain understands this a lil faster, something about an `all` that contains `map_or` confuses me.

Feel free to ignore if you disagree
```suggestion
    let earliest_expiration = metadata
        .scopes
        .iter()
        .filter_map(|scope| scope.expires_at)
        .max();
    let all_scopes_active =
        earliest_expiration.map_or(true, |expiration| current_time < expiration);
```

Reply:
no problem here, I'll take your suggestion since I have no strong preference either way.